### ISSUE TITLE
 Gibbon::MailChimpError at /visitors  the server responded with status 404

### ISSUE THREAD

**Issue by Nuttdguy** [RailsApps/learn-rails#68]
I'm getting this 404 error. I believe it means page not found?

I have installed Gibbon 2.0.0 gem. 
I have copied updated code from git to model/visitors.rb. Is this error being caused by the mailchimp_api_key not being picked up?

class Visitor
  include ActiveModel::Model
  attr_accessor :email, :string
  validates_presence_of :email
  validates_format_of :email, with: /\A[-a-z0-9_+\.]+\@([-a-z0-9]+\.)+[a-z0-9]{2,4}\z/i

  def subscribe
    mailchimp = Gibbon::Request.new(api_key: Rails.application.secrets.mailchimp_api_key)
    list_id = Rails.application.secrets.mailchimp_list_id
    result = mailchimp.lists(list_id).members.create(
      body: {
        email_address: self.email,
        status: 'subscribed'
    })
    Rails.logger.info("Subscribed #{self.email} to MailChimp") if result
  end

end

Is there a way to input to api_key directly to see whether its the local ENV variable that's causing the error?

Thanks

**Comment by patsiren** [comment]
The Gibbon gem has been updated to use mailchimp API 3.0 so some of the calls are a little different.  Check out the project directory here for specifics: https://github.com/amro/gibbon

Most notably your list request is outdated.  The new syntax is here:

gibbon.lists(list_id).retrieve

so you might try*:

gibbon.lists(Rails.application.secrets.mailchimp_lis_id).retrieve.


*take this with a grain of salt, I'm only in here because the google drive API that we're using has been officially deprecated (non-oAuth calls now return a 404), so I generally assume the rails_apps tutorials are starting to show their age.  When in doubt with any tutorial, make sure to check the gem/API docs.

**Comment by Nuttdguy** [comment]
@patsiren - I copied exactly the update / change that was implemented recently by Daniel. I too think the tutorial is dated because Google Drive and MailChimp aspects of the book do not seem to work without errors. I likely that all the updates and changes by implementing third-party gems makes it difficult to maintain.

@DanielKehoe - The updates returns but seems to return an error on the create method ::

tarted POST "/visitors" for 127.0.0.1 at 2015-09-05 20:30:13 -0500
Processing by VisitorsController#create as HTML
  Parameters: {"utf8"=>"✓", "authenticity_token"=>"EJSthNpYN8bFXq+670FEbjihsd4HymHKLrQlMo7w6a7WmHkHA3CMuzo5UhDNviyzMX2fMz1UFUWAgWEf+1KUvA==", "visitor"=>{"email"=>"myemail@yahoo.com"}, "commit"=>"Sign up for the newsletter"}
Completed 500 Internal Server Error in 281ms

Gibbon::MailChimpError - the server responded with status 404:
  gibbon (2.0.0) lib/gibbon/api_request.rb:100:in `handle_error'
  gibbon (2.0.0) lib/gibbon/api_request.rb:16:in `rescue in post'
  gibbon (2.0.0) lib/gibbon/api_request.rb:10:in `post'
  gibbon (2.0.0) lib/gibbon/request.rb:28:in `create'
  app/models/visitor.rb:10:in `subscribe'
  app/controllers/visitors_controller.rb:10:in `create'


In another attempt, I made changes to the visitor.rb ::

 def subscribe
    mailchimp = Gibbon::Request.new(api_key: Rails.application.secrets.mailchimp_api_key)
    list_id = mailchimp.lists(Rails.application.secrets.mailchimp_list_id).retrieve
    result = mailchimp.lists(list_id).members.create(
      body: {
        email_address: self.email,
        status: 'subscribed'
    })
    Rails.logger.info("Subscribed #{self.email} to MailChimp") if result
  end
end

AND was returned ::
""" bad URI(is not URI?): https://us11.api.mailchimp.com/3.0/lists/{"lists"=> """
 
The error resulting from || result = mailchimp.lists(list_id).members.create || It seems when it executes the status: subscribed line, an error is thrown because of an invalid URI. For some reason it does not add the list_id to the URI.... Is that the problem?

**Comment by DanielKehoe** [comment]
So @patsiren: It seems you are working from an outdated version of the book. Be sure you have book v2.2.0. I've removed the Google Drive chapter because there is no longer an easy way (for beginners) to authenticate with Google Drive. The correct code for Gibbon 2.0 is here in the repository (recently updated) and the next version of the book will contain the updated Gibbon 2.0 code. @Nuttdguy has the right code for Gibbon 2.0. Yes, it is difficult to keep the tutorials updated as Rails and gems and services like Google Drive change. However, it is unfair to say the tutorials are "starting to show their age." It takes time to update the tutorials but I actively maintain them.

@Nuttdguy to address the issue you opened: The status 404 error is returned when you attempt to enter an email address that is already in the MailChimp database. Is that perhaps the cause for your error?

**Comment by Nuttdguy** [comment]
@DanielKehoe -- Thanks for reply. The problem seemed to be;

""" list_id = mailchimp.lists(Rails.application.secrets.mailchimp_list_id).retrieve """
was returning results properly, but on the following line;
""" result = mailchimp.lists(list_id).members.create {body: ( ... )} ""
was not retrieving the list_id value from the first "list_id" assignment. 

So instead by changing the result line to;
""" result = mailchimp.lists(Rails.application.secrets.mailchimp_list_id).members.create( body: { ... }) ""
the code worked and email was added to the MailChimp list. 

I know you mentioned a way to debug, but is there a way to input code into the CLI to test whether parts of the code work, (i.e. that block of code specifically). I want to check the value of list_id to see whether its working; which seems that it is. So why is it that I have to call it specifically again??? Pretty confusing stuff.

**Comment by DanielKehoe** [comment]
You can use ```$ rails console``` to test the code from the CLI.

Thanks for sharing the issue. I'll investigate further.

**Comment by Nuttdguy** [comment]
Sounds good, thanks!

**Comment by Nuttdguy** [comment]
@DanielKehoe || Used the Rails Console in attempt to retrieve ENV from Rails.application.secrets and the result was this:

Loading development environment (Rails 4.2.0)
2.2.0 :001 > k = Rails.application.secrets.mailchimp_api_key
 => nil 
2.2.0 :002 > n = Rails.application.secrets.gmail_username
 => "myemail@gmail.com" 
2.2.0 :003 > Rails.application.secrets
 => {:gmail_username=>"myemail@gmail.com", :gmail_password=>"my_password", :mailchimp_api_key=>nil, :mailchimp_list_id=>nil, :domain_name=>nil, :owner_email=>"myemail@gmail.com", :secret_key_base=>"my_secret_key", :secret_token=>nil} 

Now, why would it be picking up my ENV variables for gmail_username, gmail_password BUT NOT mailchimp_api_key or mailchimp_list_id?

Below is from my .bashrc:: (Actual configuration parameters are replaced with placeholders). 

# Learn-Rails Environment Variables Additions
export GMAIL_USERNAME="myemail@gmail;.com"
export GMAIL_PASSWORD="mypassword"
export MAILCHIMP_API_KEY="API-Key-Number"
export MAILCHIMP_LIST_ID="List-id-number"
export OWNER_EMAIL="myemail@gmail.com"

The reason I ask is the configuration which was thought correct is no longer working since I've pushed it up into Heroku.

**Comment by DanielKehoe** [comment]
Did you say you're using Figaro? Remove it and what happens?

**Comment by Nuttdguy** [comment]
@DanielKehoe || Thanks for your reply. I actually got it to work now. 

After using the rails console, I was able to view the values from the secrets.yml file. For some reason, it was not picking up the configured values so I decided to manually hard code it. I went back to the console to see if the values existed and they did. 
Afterwards, I went to try and subscribe with the hard coded values in the secrets.yml variables and it was successful. 

I  replaced;
  mailchimp_api_key: <%= ENV['MAILCHIMP_API_KEY'] %>
  mailchimp_list_id: <%= ENV['MAILCHIMP_LIST_ID'] %>
 the original double quotes with single quotes, and copied the line of code from gmail_username that was working / returning a value. I returned to the console and it was working. The keys were returning values. 

So long story short, it seems that replacing the double quote and both of the lines <%= ENV['MAILCHIMP_LIST_ID'] %> solved the problem. It now works on both development and production.

**Comment by DanielKehoe** [comment]
Glad to hear you resolved the issue. Perhaps there were characters in the MAILCHIMP_API_KEY and MAILCHIMP_LIST_ID that were causing interpolation with double quotes. Reopen the issue if you have any further trouble.

**Comment by AnnaRedziak** [comment]
I came across a very similar problem, although everything works good when I run _rails s_ and open my app locally, I can send e-mails via contact and I can subscribe. When I want to open the app with herokuapp.com I'm getting

>  Gibbon::MailChimpError (the server responded with status 404 @title="Resource Not Found", @detail="The requested resource could not be found."

no subscribing works, no sending e-mails via contact page. Any idea what could go wrong?
